### ISSUE TITLE
(Linux) ivpn login & logout always disables firewall

### ISSUE THREAD

**Issue by Enteee** [ivpn/desktop-app#87]
# Bug report

## Describe your environment

* Device: Laptop
* OS name and version: fedora-33  
* IVPN app version: 3.3.30

## Describe the problem

On login a previously enabled firewall is always disabled. This even happens when the firewall is configured to be always on.

### Steps to reproduce:

1. Logout: `ivpn logout`
2. Enable firewall: `ivpn firewall -on`
3. Check that firewall is enabled: `ivpn firewall` -> Firwall is enabled, iptables rules present
4. login: `ivpn login .....`
5. Check firewall again -> Firewall is now disabled, iptables rules are removed.

### Observed Results:

* Login resets iptables rules (disables firewall)

### Expected Results:

* Login should preserve firewall state

### Relevant Code:

```
ivpn logout
ivpn firewall -on
ivpn login .....
ivpn firewall
```

### Background

I am currently evaluating ivpn to be deployed onto managed clients. A crucial requirement is the functionality of the fail-close or "always on" firewall. This means, under **absolutely no circumstances** there must be any traffic leaking.

For that I do deploy ivpn with the `IsFwPersistant` value in `/opt/ivpn/mutable/settings.json` changed to `true`. When I then start `ivpn-service` the firewall is set up as expected. But as soon as a user logs in, all firewall rules are truncated. Which is really bad.

**Comment by Enteee** [comment]
I did some more investigation on this topic and it seems that the "always on" firewall feature seems to be completely broken.

The firewall also is reset when I logout.

Reproduce:

1. `ivpn login .....`
2. In the ui enable Always-on firewall
3. Check that firewall is enabled: `ivpn firewall` -> Firwall is enabled, iptables rules present
4. `ivpn logout`
5. Check firewall again -> Firewall is now disabled, iptables rules are removed.

**Comment by stenya** [comment]
Hi @Enteee ,

The IVPN Firewall functionality works as designed.

Please, do not perform manual configuration of the internal IVPN files.
The 'settings.json' file is intended to be changed only by the IVPN daemon.

Commands `ivpn login` and `ivpn logout` are disabling Firewall. It is intended behavior.
It is required to avoid any problems with connectivity when a user just installed the IVPN app (or going to uninstall it).

I see there is only one potential change that can be implemented: IVPN CLI has to not allow to use of Firewall commands until a user logged in. This will avoid misunderstanding of Firewall functionality behavior.

**Comment by Enteee** [comment]
Hello @stenya  and thank you for your reply.

> The IVPN Firewall functionality works as designed.

From a feature called "Always-On Firewall" which advertises itself with:

> [...] IVPN Firewall will always be active even  when IVPN Client is not running.

I would expect that it is actually "Always-On"... And does not disable itself when I log out / in.

I even see this as a possible security issue in the following scenario:

1. I am logged in on device A and enable always-on firewall.
2. Somebody tricks me into a login on device B (or I do this accidentally).
3. Because I only have a single use license, I confirm logout on all devices.
4. Device A logs out and disabled Firewall.
5. Device A now starts leaking traffic.

This would be bad bad.

I found that when i juggle a bit with the order of the inserts to your `OUT_IVPN` and `IN_IVPN` chains I can ensure API connectivity even when the user is logged out and always on firewall is enabled. I fully agree with you that the always-on firewall should be disabled during uninstall. But that should be the task of the uninstall routine.

On the subject of changing `settings.json`:
I really think your app should support changing of default configuration. To be honest, I am not quite sure how your app can pass any professional evaluation without. Your shipped default configuration will never fit all requirements and environments - no matter what you do.

For example,
I need the always-on firewall to be on by default.
And I need the anti-tracking feature be enabled by default.
Because If I don't enable those things by default, only a minority of my users ever will.

Look, I like your app. Its small, open, and does what It is supposed to do. I am happy to contribute with input and maybe even with small code changes. its just that when the second sentence of a developer response is "works as designed" I start considering other options...

**Comment by stenya** [comment]
@Enteee we are trying to find a solution for your requirements.
We do not want to change the current behavior.

There is an idea to extend CLI commands `ivpn login` and `ivpn logout` by an additional optional argument (something like `-keep_firewall_state`).
In this case, the firewall state will stay unchanged when you do commands login\logout from CLI. In the logged-out state (even if the firewall is enabled) there will be still the possibility to access the IVPN API server.
What do you think?

**Comment by cipheromega** [comment]
@Enteee These are good points and I agree that we should default to saving firewall state. The always-on firewall was developed before we implemented login/logout sessions. We will implement saving of app configuration on logout as default in the next release.

**Comment by gorkapernas** [comment]
This has been implemented and tested on https://github.com/ivpn/desktop-app/issues/91

The possibility to configure the firewall status while logged out has been added to the CLI as well as to the GUI (the switcher to enable/disable the Firewall will be found in the login screen), it will also be possible to keep all application settings on log out, not only the Firewall status, however, the firewall will be handled separately from the rest of the settings when logging out. The user can also choose to reset everything to default during the log out.

Please see https://github.com/ivpn/desktop-app/issues/91 for further details.

**Comment by stenya** [comment]
Released in v3.4.0

**Comment by Enteee** [comment]
Thank you for implementing and releasing this. @stenya , @gorkapernas 
I will review this in the next days and provide feedback accordingly.
